Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,979 @@
+
+```python
+# Using a lambda expression to define a function
+# "on the fly"
+even_integers = st.integers().map(lambda x: 2 * x)
+```
+
+```python
+>>> even_integers.example()
+220
+```
+
+In general, the syntax for defining a lambda expression is:
+
+```
+lambda <comma-separated variable names>: <expression using variables>
+```
+
+Note that lambda expressions are restricted compared to typical function definitions: their body must consist only of a single Python expression, and the lambda function returns whatever that expression returns.
+
+Here are some examples of lambda expressions:
+
+```python
+# function definition
+def add(x, y):
+    return x + y
+
+>>> add(2, 3)
+5
+
+# equivalent lambda expression
+>>> (lambda x, y: x + y)(2, 3)
+5
+
+# function definition
+def get_first_and_last_items(x):
+    return x[0], x[-1]
+
+>>> get_first_and_last_items(range(11))
+(0, 10)
+
+# equivalent lambda expression
+>>> (lambda x: x[0], x[-1])(range(11))
+(0, 10)
+```
+
+We will make keen use of lambdas in order to enrich our Hypothesis strategies. 
+<!-- #endregion -->
+
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Using the `.map` method to create a sorted list**
+
+Using the `.map()` method, construct a Hypothesis strategy that produces a sorted list of integers.
+Generate some examples from your strategy and check that they are sorted (we may have to generate quite a few examples to see a diverse set of values)
+
+</div>
+
+
+<div class="alert alert-info"> 
+
+**Reading Comprehension: Getting creative with the `.map` method**
+
+Construct a Hypothesis strategy that produces either the string `"cat"` or the string `"dog"`.
+Then, write a test that uses this strategy;
+it should simply test that either `"cat"` or `"dog"` was indeed produced by the strategy.
+Run the test.
+
+</div>
+
+
+<!-- #region -->
+### The `.filter` method
+
+Hypothesis strategies can also [have their data filtered](https://hypothesis.readthedocs.io/en/latest/data.html#filtering) via the `.filter` method. 
+`.filter` takes a function (or any "callable") that accepts as input the data generated by the strategy, and returns:
+
+ - `True` if the data should pass through the filter
+ - `False` if the data should be rejected by the filter
+
+Consider, for instance, that you want to generate all integers other than `0`.
+You can write the filtered strategy:
+
+```python
+# Demonstrating the `.filter()` method
+non_zero_integers = st.integers().filter(lambda x: x != 0)
+```
+
+The `.filter` method is not magic – it is not able to "just know" how to avoid generating all barred values. 
+A strategy that filters our too much data will prompt Hypothesis to raise an error.
+For example, let's try to filter `st.integers()` so that it only produces values on $[10, 20]$.
+
+```python
+# Using `.filter()` to filter out a large proportion of generated
+# values will result in an error
+@given(st.integers().filter(lambda x: 10 <= x <= 20))
```

Comment:
Unfortunately (?), this example *does now work* thanks to our filter-rewriting logic for intervals!  I suggest using "multiples of 17" or similar.